pkg/apis/core/validation/validation.go

```diff
@@ -1501,8 +1501,7 @@ var supportedReclaimPolicy = sets.NewString(string(core.PersistentVolumeReclaimD
 
 var supportedVolumeModes = sets.NewString(string(core.PersistentVolumeBlock), string(core.PersistentVolumeFilesystem))
 
-var supportedDataSourceKinds = sets.NewString(string("VolumeSnapshot"))
-var supportedDataSourceAPIGroups = sets.NewString(string("snapshot.storage.k8s.io"))
+var supportedDataSourceAPIGroupKinds = sets.NewString(string("snapshot.storage.k8s.io:VolumeSnapshot"))
```

Comment:
this string format isn't going to make sense in an error message. I expected to see the structured types, like:
```go
allowedGroupKinds = map[schema.GroupKind]bool{
  schema.GroupKind{Group: "snapshot.storage.k8s.io", Kind: "VolumeSnapshot"}: true,
}
```

and stringified versions of those in the error message

Reply:
Does this look okay?

                groupKind := schema.GroupKind{Group: "", Kind: spec.DataSource.Kind}
                if spec.DataSource.APIGroup != nil {
                        groupKind.Group = string(*spec.DataSource.APIGroup)
                }
                groupKindList := make([]string, 0, len(supportedDataSourceAPIGroupKinds))
                for grp, _ := range supportedDataSourceAPIGroupKinds {
                        groupKindList = append(groupKindList, fmt.Sprintf("%#v", grp))
                }
                if !supportedDataSourceAPIGroupKinds[groupKind] {
                        allErrs = append(allErrs, field.NotSupported(fldPath.Child("dataSource"), groupKind, groupKindList))
                }
        }

I need to convert the map to a []string.

k8s.io/apimachinery/pkg/util/validation/field/errors.go
func NotSupported(field *Path, value interface{}, validValues []string) *Error

Reply:
that's fine, let's use the GroupKind#String() representation for the error message

Reply:
Sure.